include/exadg/operators/grid_related_time_step_restrictions.h

```diff
@@ -0,0 +1,349 @@
+  new_time_step = dealii::Utilities::MPI::min(new_time_step, mpi_comm);
+
+  // Cut time step size after, e.g., 4 digits of accuracy in order to make sure that there is no
+  // drift in the time step size depending on the number of processors when using adaptive time
+  // stepping. This effect can occur since the velocity field and the time step size are coupled
+  // (there is some form of feedback loop in case of adaptive time stepping, i.e., a minor change
+  // in the time step size due to round-off errors implies that the velocity field is evaluated
+  // at a slightly different time in the next time step and so on). This way, it can be ensured
+  // that the sequence of time step sizes is exactly reproducible with the results being
+  // independent of the number of processors, which is important for code verification in a
+  // parallel setting.
+  new_time_step = dealii::Utilities::truncate_to_n_digits(new_time_step, 4);
+
+  return new_time_step;
+}
+
+/*
+ * this function computes the actual CFL number in each cell given a global time step size
+ * (that holds for all cells)
+ */
+
+template<int dim, typename value_type>
+void
+calculate_cfl(dealii::LinearAlgebra::distributed::Vector<value_type> &       cfl,
+              dealii::Triangulation<dim> const &                             triangulation,
+              dealii::MatrixFree<dim, value_type> const &                    data,
+              unsigned int const                                             dof_index,
+              unsigned int const                                             quad_index,
+              dealii::LinearAlgebra::distributed::Vector<value_type> const & velocity,
+              double const                                                   time_step,
+              unsigned int const                                             degree,
+              double const                                                   exponent_fe_degree)
+{
+  CellIntegrator<dim, dim, value_type> fe_eval(data, dof_index, quad_index);
+
+  unsigned int const n_active_cells = triangulation.n_active_cells();
+
+  cfl.reinit(n_active_cells);
+
+  for(unsigned int cell = 0; cell < data.n_cell_batches(); ++cell)
+  {
+    fe_eval.reinit(cell);
+    fe_eval.read_dof_values(velocity);
+    fe_eval.evaluate(dealii::EvaluationFlags::values);
+
+    dealii::VectorizedArray<value_type> u_va =
+      dealii::make_vectorized_array<value_type>(std::numeric_limits<value_type>::min());
+
+    dealii::Tensor<2, dim, dealii::VectorizedArray<value_type>> invJ;
+    dealii::Tensor<1, dim, dealii::VectorizedArray<value_type>> u_x;
+    dealii::Tensor<1, dim, dealii::VectorizedArray<value_type>> ut_xi;
+
+    // loop over quadrature points
+    for(unsigned int q = 0; q < fe_eval.n_q_points; ++q)
+    {
+      u_x   = fe_eval.get_value(q);
+      invJ  = fe_eval.inverse_jacobian(q);
+      invJ  = transpose(invJ);
+      ut_xi = invJ * u_x;
+
+      u_va = std::max(u_va, ut_xi.norm());
```

Comment:
I am not so sure about this type of change, since it probably makes the code harder to read/understand. The question is whether this change is really that relevant regarding the overall performance. If we make the change, the variable names should be changed to `_square` as well.